File: samples/java/SupportChat/Shared/src/main/java/systems/zlink/samples/supportchat/shared/contracts/Messages.java
package systems.zlink.samples.supportchat.shared.contracts;

import systems.zlink.framework.actors.ActorRefSnapshot;

public final class Messages {
    private Messages() {}

    public record AuthenticateReq(String accessToken) {}

    public record AuthenticateRes(String actorId, String displayName, String role) {}

    public record AuthenticateUserReq(String accessToken) {}

    public record AuthenticateUserRes(
            boolean accepted, String actorId, String displayName, String role, String reason) {}

    public record OpenConversationApiReq(
            String customerActorId, String customerDisplayName, String subject) {}

    public record OpenConversationApiRes(String conversationId, String status) {}

    public record AllocateConversationReq(
            String customerActorId, String customerDisplayName, String subject) {}

    public record AllocateConversationRes(String conversationId, String status) {}

    public record EnsureSupportUserActorReq(
            String actorId, String displayName, String role, String participantId) {}

    public record EnsureSupportUserActorRes(ActorRefSnapshot actor) {}

    public record EnsureAgentConversationReq(
            String rosterActorId, String displayName, String conversationId) {}

    public record EnsureAgentConversationRes(ActorRefSnapshot actor) {}

    public record OpenConversationReq(String subject) {}

    public record OpenConversationRes(String conversationId, ConversationState state) {}

    public record SetAgentAvailableReq(boolean isAvailable) {}

    public record SetAgentAvailableRes(boolean isAvailable) {}

    public record JoinConversationReq(
            String conversationId, String participantId, String role, String displayName) {}

    public record JoinConversationRes(boolean scheduled, String actorId, ConversationState state) {}

    public record JoinConversationFailedNotify(
            String conversationId, String error, boolean isRetriable) {}

    public record SendChatMessageReq(String text) {}

    public record SendChatMessageRes(ChatMessage message, ConversationState state) {}

    public record SetTypingMsg(boolean isTyping) {}

    public record CloseConversationReq(String reason) {}

    public record CloseConversationRes(ConversationState state) {}

    public record ParticipantJoinedNotify(
            String conversationId, String actorId, String role, ConversationState state) {}

    public record ConversationAssignedNotify(String conversationId, ConversationState state) {}

    public record ChatMessageNotify(
            String conversationId, ChatMessage message, ConversationState state) {}

    public record TypingChangedNotify(
            String conversationId, String actorId, boolean isTyping, ConversationState state) {}

    public record ConversationIdleNotify(String conversationId, ConversationState state) {}

    public record ConversationClosedNotify(String conversationId, ConversationState state) {}

    public record ConversationState(
            String conversationId,
            String subject,
            String status,
            String customerActorId,
            String agentActorId,
            long lastMessageSeq,
            Long lastMessageAtUnixMs,
            Long idleDeadlineUnixMs) {}

    public record ChatMessage(
            String conversationId,
            long messageSeq,
            String senderActorId,
            String text,
            long sentAtUnixMs) {}
}

File: samples/kotlin/SupportChat/Shared/src/main/kotlin/systems/zlink/samples/kotlin/supportchat/shared/contracts/Messages.kt
package systems.zlink.samples.kotlin.supportchat.shared.contracts

import systems.zlink.framework.actors.ActorRefSnapshot

data class AuthenticateReq(val accessToken: String)

data class AuthenticateRes(val actorId: String, val displayName: String, val role: String)

data class AuthenticateUserReq(val accessToken: String)

data class AuthenticateUserRes(
    val accepted: Boolean,
    val actorId: String?,
    val displayName: String?,
    val role: String?,
    val reason: String?,
)

data class OpenConversationApiReq(
    val customerActorId: String,
    val customerDisplayName: String,
    val subject: String,
)

data class OpenConversationApiRes(val conversationId: String, val status: String)

data class AllocateConversationReq(
    val customerActorId: String,
    val customerDisplayName: String,
    val subject: String,
)

data class AllocateConversationRes(
    val conversationId: String,
    val status: String,
    val state: ConversationState? = null,
)

data class EnsureSupportUserActorReq(
    val actorId: String,
    val displayName: String,
    val role: String,
    val participantId: String = actorId,
)

data class EnsureSupportUserActorRes(val actor: ActorRefSnapshot)

data class EnsureAgentConversationReq(
    val rosterActorId: String,
    val displayName: String,
    val conversationId: String,
)

data class EnsureAgentConversationRes(
    val actor: ActorRefSnapshot,
    val scheduled: Boolean,
    val state: ConversationState,
)

data class OpenConversationReq(val subject: String)

data class OpenConversationRes(val conversationId: String, val state: ConversationState)

data class SetAgentAvailableReq(val isAvailable: Boolean)

data class SetAgentAvailableRes(val isAvailable: Boolean)

data class JoinConversationReq(
    val participantId: String = "",
    val role: String = "",
    val displayName: String = "",
)

data class JoinConversationRes(val scheduled: Boolean, val state: ConversationState)

data class JoinConversationFailedNotify(
    val conversationId: String,
    val error: String,
    val isRetriable: Boolean,
)

data class SendChatMessageReq(val text: String)

data class SendChatMessageRes(val message: ChatMessage, val state: ConversationState)

data class SetTypingMsg(val isTyping: Boolean)

data class CloseConversationReq(val reason: String?)

data class CloseConversationRes(val state: ConversationState)

data class ParticipantJoinedNotify(
    val conversationId: String,
    val actorId: String,
    val role: String,
    val state: ConversationState,
)

data class ConversationAssignedNotify(val conversationId: String, val state: ConversationState)

data class ChatMessageNotify(
    val conversationId: String,
    val message: ChatMessage,
    val state: ConversationState,
)

data class TypingChangedNotify(
    val conversationId: String,
    val actorId: String,
    val isTyping: Boolean,
    val state: ConversationState,
)

data class ConversationIdleNotify(val conversationId: String, val state: ConversationState)

data class ConversationClosedNotify(val conversationId: String, val state: ConversationState)

data class ConversationState(
    val conversationId: String,
    val subject: String,
    val status: String,
    val customerActorId: String,
    val agentActorId: String?,
    val lastMessageSeq: Long,
    val lastMessageAtUnixMs: Long?,
    val idleDeadlineUnixMs: Long?,
)

data class ChatMessage(
    val conversationId: String,
    val messageSeq: Long,
    val senderActorId: String,
    val text: String,
    val sentAtUnixMs: Long,
)
